### PR TITLE
Socketselect

### PR DESCRIPTION
**Pull request by kapkic**
Cross PR w/ Nick.
Added socketselect.c and .py test case.
Polished suite: Different tests are more noticeable. Failed tests appear in the end as a list to help debugging.

### CODE REVIEW

File: tests/test_cases/socketselect.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    exit(-1)
+
+resultLen = len(lind_result_split)
+for line in range(0,resultLen):
+    if lind_result_split[line].translate(None, string.digits+'-') != native_result_split[line].translate(None, string.digits+'-'):
+        print "Nondeterministic lines of an unrecognized format!"
+        exit(-1)
+
+# Hello received check
+if lind_result_split[7] != native_result_split[7]:
```
Comment (rennergade) [inline tests/test_cases/socketselect.py:25]: is this always deterministically at this array index? I thought there could be print ordering differences.

**Review by rennergade — COMMENTED** [review]
Just one small comment. I like the additions you made to the suite!

File: tests/test_cases/socketselect.py (hunk 1 above)
Comment (kapkic) [reply tests/test_cases/socketselect.py:25]: AFAIK; print order differences happen with fork. But to make sure, I'll sort and do the check.

Comment (rennergade) [reply tests/test_cases/socketselect.py:25]: This is threaded so it should potentially have the same issues.

File: tests/test_cases/socketselect.c [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+   {
+      perror("ioctl() failed");
+      close(listen_sd);
+      exit(-1);
+   }
+
+   
+   /* Bind the socket                                           */
+   
+   memset(&addr, 0, sizeof(addr));
```
Comment (jesings) [inline tests/test_cases/socketselect.c:93]: No need to memset if you will just memcpy over that

File: tests/test_cases/socketselect.c [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+   }
+
+   
+   /* Initialize the master fd_set                              */
+   
+   FD_ZERO(&master_set);
+   max_sd = listen_sd;
+   FD_SET(listen_sd, &master_set);
+
+   /* Initialize the timeval struct to 3 minutes.  If no        */
```
Comment (jesings) [inline tests/test_cases/socketselect.c:124]: Update comment to say 5 seconds

File: tests/test_cases/socketselect.c [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                  /* server.                                    */
+                  
+                  new_sd = accept(listen_sd, NULL, NULL);
+                  if (new_sd < 0)
+                  {
+                    //  if (errno != EWOULDBLOCK)
+                    //  {
+                    //     perror("  accept() failed");
+                    //     end_server = TRUE;
+                    //  }
```
Comment (jesings) [inline tests/test_cases/socketselect.c:204]: Why is this commented out?

**Review by jesings — APPROVED** [review]
Overall looks great! 3 minor changes.

File: tests/test_cases/socketselect.c (hunk 2 above)
Comment (rennergade) [reply tests/test_cases/socketselect.c:93]: This was from some of the code I ported, fine with it being removed.

File: tests/test_cases/socketselect.c (hunk 3 above)
Comment (rennergade) [reply tests/test_cases/socketselect.c:124]: Sure.

File: tests/test_cases/socketselect.c (hunk 4 above)
Comment (rennergade) [reply tests/test_cases/socketselect.c:204]: Ah this was commented out when I was trying to get this to work despite the errnos being out of whack. Can uncomment this now.

**Comment by kapkic** [comment]
PR changes are submitted. This CI will fail as we haven't merged the Glibc accept syscall fix into develop.

**Comment by kapkic** [comment]
This check succeeded on it's fourth run.
This is the output for failed test cases:
```
Running test: socketselect.c
Lind Output:
Waiting on select()...
Hello message sent

  select() timed out.  End program.
------------------------------------------------------------------
Regular Output:
Waiting on select()...
Hello message sent
  Listening socket is readable
  New incoming connection - 6
Waiting on select()...
  Descriptor 6 is readable
  17 bytes received
Hello from client
  Connection closed
Waiting on select()...
  select() timed out.  End program.
```

**Review by rennergade — APPROVED** [review]

**Pull request merged**